autosklearn/util/data.py

```diff
@@ -70,3 +132,223 @@ def predict_RAM_usage(X: np.ndarray, categorical: List[bool]) -> float:
             estimated_columns += 1
     estimated_ram = estimated_columns * X.shape[0] * X.dtype.itemsize
     return estimated_ram
+
+
+def subsample(
+    X: SUPPORTED_FEAT_TYPES,
+    y: Union[List, np.ndarray, pd.DataFrame, pd.Series],
+    is_classification: bool,
+    sample_size: Union[float, int],
+    random_state: Optional[Union[int, np.random.RandomState]] = None,
+) -> Tuple[SUPPORTED_FEAT_TYPES, Union[List, np.ndarray, pd.DataFrame, pd.Series]]:
+    """ Subsamples data returning the same type as it recieved.
+
+    If `is_classification`, we split using a stratified shuffle split which
+    preserves unique labels in the training set.
+
+    NOTE:
+    It's highly unadvisable to use lists here. In order to preserver types,
+    we convert to a numpy array and then back to a list.
+
+    NOTE2:
+    Interestingly enough, StratifiedShuffleSplut and descendants don't support
+    sparse `y` in `split(): _check_array` call. Hence, neither do we.
+
+    NOTE3:
+    The core autosklearn library doesn't rely on the full type of X.
+    The typing could be reduced to:
+    *   X: np.ndarray | spmatrix
+    *   Y: np.ndarray
+
+
+    Parameters
+    ----------
+    X: SUPPORTED_FEAT_TYPES
+        The X's to subsample
+
+    Y: List | np.ndarray | pd.DataFrame | Series
+        The Y's to subsample
+
+    is_classification: bool
+        Whether this is classification data or regression data. Required for
+        knowing how to split.
+
+    sample_size: float | int
+        If float, percentage of data to take otherwise if int, an absolute
+        count of samples to take.
+
+    random_state: int | RandomState = None
+        The random state to pass to the splitted
+
+    Returns
+    -------
+    (SUPPORTED_FEAT_TYPES, List | np.ndarray | pd.DataFrame | Series)
+        The X and y subsampled according to sample_size
+    """
+    if isinstance(X, list):
+        X = np.asarray(X)
+
+    if isinstance(y, list):
+        y = np.asarray(y)
+
+    if is_classification:
```

Comment:
I'm getting a bit worried that this might not work for multi-label classification?

Reply:
I don't see why it wouldn't? The only logic that directly manipulates `y` is converting it to an `np.ndarray` if it's a list.

I looked into the `CustomStratifiedShuffleSplit` and it seems to work with all of them as well, where again, the only manipulation we do is by indexing a row. We don't care if the index points to a row `[1,0,1]` in multi-label or `4` if it's multi-class.

Reply:
True, learned something new today :)

https://github.com/scikit-learn/scikit-learn/blob/0d378913b/sklearn/model_selection/_split.py#L1930